### PR TITLE
Use the Knp menu for the back end header menu

### PR DESCRIPTION
**Pull request by leofeyer**
This PR makes the necessary changes so the back end header menu is built via Knp menu and the debug button can be added in the manager bundle (this was an open TODO).

The PR also "fixes" the back end main menu, which does not adhere to the Knp menu standards and thus is only partially extendable.

### TODOs

- [x] Rework the main menu
- [x] Fix the existing unit tests
- [x] Add unit tests for the new classes

### CODE REVIEW

File: core-bundle/src/Resources/contao/languages/en/modules.xlf
Comment (aschempp) [inline core-bundle/src/Resources/contao/languages/en/modules.xlf]: how is this related to the PR?

**Review by aschempp — CHANGES_REQUESTED** [review]
I think we should keep the back end menu naming as-is, meaning the `be_menu.twig` and the respective event. And use a new template and a new event name for the `be_header_menu`, as devs listening to the `BACKEND_MENU_BUILD` event are for sure not checking the root node name atm.

File: core-bundle/src/Resources/contao/languages/en/modules.xlf
Comment (leofeyer) [reply core-bundle/src/Resources/contao/languages/en/modules.xlf]: It is not. Reverted in d0b9a00000ee703d54bdc38d91ff6efcd6e37fdd.

**Comment by leofeyer** [comment]
I guess you are right, although it is ugly to have two different events for basically the same thing in the same context. But changed in d0b9a00000ee703d54bdc38d91ff6efcd6e37fdd now.

File: core-bundle/src/Event/ContaoCoreEvents.php
Comment (aschempp) [inline core-bundle/src/Event/ContaoCoreEvents.php]: As this is for Contao 4.9 only, we should actually not use an event name, but a different event class, shouldn't we? If not, I would still suggest to simply name this `BACKEND_HEADER_BUILD` (and shortname accordingly)

File: core-bundle/src/EventListener/BackendMenuListener.php
Comment (aschempp) [inline core-bundle/src/EventListener/BackendMenuListener.php]: Previously, there was a separation between the system message count and the error count. Can you explain why that's no longer the case?

File: core-bundle/src/Event/ContaoCoreEvents.php
Comment (leofeyer) [reply core-bundle/src/Event/ContaoCoreEvents.php]: You suggestion shows exactly why using two different events is not good in this case. 😄 It would simply be the `BackendMenuBuilder::class` event if we were able to use the same event in both methods.

But since we cannot, I guess we have to stick with `BACKEND_HEADER_MENU_BUILD`. Because `BACKEND_HEADER_BUILD` is misleading – after all, the event is triggered only when the back end header menu is build, not when the back end header is build.

File: core-bundle/src/EventListener/BackendMenuListener.php
Comment (leofeyer) [reply core-bundle/src/EventListener/BackendMenuListener.php]: Because `$systemMessagesCount` is not used anywhere (not even in Contao 4.8). Seemed like a left-over, therefore I have not added it in the new class.

Comment (aschempp) [reply core-bundle/src/EventListener/BackendMenuListener.php]: And we're not showing non-error messages anymore either?

File: core-bundle/src/Event/ContaoCoreEvents.php
Comment (aschempp) [reply core-bundle/src/Event/ContaoCoreEvents.php]: > You suggestion shows exactly why using two different events is not good in this case. 😄 It would simply be the BackendMenuBuilder::class event if we were able to use the same event in both methods.

I know we can use the same event, but we don't have to 😉 

> But since we cannot, I guess we have to stick with BACKEND_HEADER_MENU_BUILD. Because BACKEND_HEADER_BUILD is misleading – after all, the event is triggered only when the back end header menu is build, not when the back end header is build.

I don't find it misleading. It's a KnpMenu Event, so it's obviously about the menu regardless of the event name. But either is fine with me 😊

File: core-bundle/src/EventListener/BackendMenuListener.php
Comment (leofeyer) [reply core-bundle/src/EventListener/BackendMenuListener.php]: We are showing them in the popup window, but the counter of the alerts button only shows error messages. This is the current behaviour in Contao 4.4+.

https://github.com/contao/contao/blob/4.4/core-bundle/src/Resources/contao/templates/backend/be_main.html5

**Comment by leofeyer** [comment]
I am still not convinced that using two events is the right thing to do. Also, I doubt that it would cause any problems, because extensions listening to the `BACKEND_HEADER_BUILD` event in Contao 4.8 (e.g. the manager bundle), will do something like this:

```php
$categoryNode = $event->getTree()->getChild('system');

if (null === $categoryNode) {
    return;
}
```

And since there is no intersection of child nodes, this should never cause any problems:

**Main menu**

 - content
 - design
 - accounts
 - system

**Header menu**

 - alerts
 - preview
 - submenu
 - burger

@contao/developers What do you think? Can we use the same event for both menus?

**Comment by ausi** [comment]
If I can differentiate between header and main navigation easily (like `$event->getTree()->getName() === 'headerMenu'`?) and it doesn’t break existing code I’m in favor of one event.

**Comment by Toflar** [comment]
I'm with @ausi except for the fact that I would like to compare to a class constant rather than a regular string.

**Comment by leofeyer** [comment]
> like `$event->getTree()->getName() === 'headerMenu'`?

Exactly like this. But in Yoda style. 😂

File: core-bundle/src/EventListener/BackendMenuListener.php
Comment (aschempp) [reply core-bundle/src/EventListener/BackendMenuListener.php]: But if there is no count on error messages, we do not generate the menu item at all? I think that was also not the case previously, because we always showed the popup link?

**Review by aschempp — COMMENTED** [review]
The changes in the `config.php` files across all bundles are also not related to this PR at all. Maybe you can create a separate one (according to what we just discussed yesterday) to change that across all bundles. I'm not sure about the implications, but we should discuss that in the separate PR.

File: core-bundle/src/EventListener/BackendMenuListener.php
Comment (leofeyer) [reply core-bundle/src/EventListener/BackendMenuListener.php]: Of course we still show the menu item; it is always added in the `BackendMenuBuilder`. The listener just enhances the menu by adding the system error messages count.

**Comment by leofeyer** [comment]
Of course they are related to this PR. The PR reworks the back end menus and the `config.php` changes affect the menu item order.

File: core-bundle/src/EventListener/BackendMenuListener.php
Comment (Toflar) [inline core-bundle/src/EventListener/BackendMenuListener.php]: Was that listener introduced in 4.9? If not, then this is theoretically a BC break, as is the additional argument in the constructor. I know we said listeners aren't really API but as long as it's technically possible, people will do it.

File: core-bundle/src/Menu/BackendMenuBuilder.php
Comment (Toflar) [inline core-bundle/src/Menu/BackendMenuBuilder.php]: Same here, these changes are breaking BC imho. Again, I don't know why anyone would have ever used that class directly but as it is not `final` it would have been technically possible.

File: core-bundle/src/Routing/PreviewUrlGenerator.php
Comment (Toflar) [inline core-bundle/src/Routing/PreviewUrlGenerator.php]: Is that really an issue here? Maybe we should allow to return `null` and in that case, the URL is just not added to the header menu?

**Review by Toflar — COMMENTED** [review]
Neat work here! Well done! I have my concerns regarding BC breaks and added them as comments so we can discuss. Otherwise this PR is in a good shape 👍

File: core-bundle/src/EventListener/BackendMenuListener.php
Comment (aschempp) [reply core-bundle/src/EventListener/BackendMenuListener.php]: well the listener is a complete rewrite and has very little to do with the previous one. But we never really discussed `final` and whether we should use it, so I don't think this should be in this PR.

File: core-bundle/src/EventListener/BackendMenuListener.php
Comment (aschempp) [reply core-bundle/src/EventListener/BackendMenuListener.php]: Can you please explain the code to me? If `$count` is smaller than `1`, the method returns. Otherwise it does `$factory->createItem()`. Isnt' that `$factory->createItem()` related to adding the notification item at all?

PS: please don't mark _my_ questions as resolved, if _you_ answer them. I think only the original author should do that, except if e.g. you commit a change that solves the question.

File: core-bundle/src/Menu/BackendMenuBuilder.php
Comment (m-vo) [reply core-bundle/src/Menu/BackendMenuBuilder.php]: OT: Imho including constructor signatures in the BC promise doesn't work well with DI/composition and is only ever a problem when also utilizing inheritance. Adding `final` everywhere imo is not yet practical. Can we maybe tag classes like this with an annotation in the future? Or the constructors with `@internal`?

File: core-bundle/src/Routing/PreviewUrlGenerator.php
Comment (leofeyer) [reply core-bundle/src/Routing/PreviewUrlGenerator.php]: This is the default exception that we throw everywhere where we retrieve the current request from the request stack. It should never be thrown in real life.

Comment (Toflar) [reply core-bundle/src/Routing/PreviewUrlGenerator.php]: Ok.

File: core-bundle/src/Menu/BackendMenuBuilder.php
Comment (leofeyer) [reply core-bundle/src/Menu/BackendMenuBuilder.php]: I am with @m-vo on this. Constructor arguments are not covered by the BC promise, because that would make DI unnecessarily complicated. Does Symfony guarantee BC for constructors?

File: core-bundle/src/EventListener/BackendMenuListener.php
Comment (leofeyer) [reply core-bundle/src/EventListener/BackendMenuListener.php]: The same listener exists in the manager bundle and the class is final there (also in Contao 4.8). I guess I have copied it from there.

But @aschempp is right, we should not use `final` until we have discussed the concept. I will therefore adjust the two classes in the manager bundle as well.

File: core-bundle/src/Menu/BackendMenuBuilder.php
Comment (Toflar) [reply core-bundle/src/Menu/BackendMenuBuilder.php]: Jep, it does: https://symfony.com/doc/current/contributing/code/bc.html
Adding arguments to public methods is not BC and `public function __construct()` of course is a public method. Same goes for adding `final`.

File: core-bundle/src/EventListener/BackendMenuListener.php
Comment (leofeyer) [reply core-bundle/src/EventListener/BackendMenuListener.php]: > If `$count` is smaller than `1`, the method returns. Otherwise it does `$factory->createItem()`

If `$count` is smaller than `1`, the method returns. Otherwise it does `$item->setLabel()` and not `$factory->createItem()`! You should review more carefully.

File: core-bundle/src/Menu/BackendMenuBuilder.php
Comment (leofeyer) [reply core-bundle/src/Menu/BackendMenuBuilder.php]: So you want me to make all arguments optional and throw a lot of exceptions instead?

File: core-bundle/src/EventListener/BackendMenuListener.php
Comment (leofeyer) [reply core-bundle/src/EventListener/BackendMenuListener.php]: Fixed in 6f2250625f7325b1cf43119c356a5a7517102499.

File: core-bundle/src/Menu/BackendMenuBuilder.php
Comment (Toflar) [reply core-bundle/src/Menu/BackendMenuBuilder.php]: Throwing exceptions would be just the same, it would be a BC break. BC would mean that things still work the same they did before.
But if I look at the `BackendMenuBuilder` I ask myself why it builds the previewUrl, the logoutUrl and then dispatches the event? Building these two menu items and adding them to the menu could be done in their separate listeners just like any other menu item too? Then they are new services with their own arguments and be perfectly BC.
I mean, for the previewUrl you don't need the security service and you don't need the current request. You do for the logout listener and switch user stuff, though. So that kind of smells like you're mixing responsibilities.

Comment (leofeyer) [reply core-bundle/src/Menu/BackendMenuBuilder.php]: It is a menu builder. Its task is to build the menu. Not to dispatch an event and use five different listeners to add the menu items.

Comment (Toflar) [reply core-bundle/src/Menu/BackendMenuBuilder.php]: Well, I think it's exactly the responsibility. The builder doesn't need all of the services you're currently injecting. The ones that would like to add the navigation items do.

Comment (leofeyer) [reply core-bundle/src/Menu/BackendMenuBuilder.php]: Ok, feel free to refactor this. 🤷‍♂

**Review by Toflar — CHANGES_REQUESTED** [review]
As stated in inline comments, this currently introduces unnessary BC breaks which could be nicely solved using separate listeners for the menu items, just like it's done e.g. for the debug link in the manager-bundle.

**Comment by leofeyer** [comment]
> I ask myself why it builds the previewUrl, the logoutUrl and then dispatches the event?

The menu builder does not build these URLs – the injected generators do.

> BC breaks which could be nicely solved using separate listeners for the menu items, just like it's done e.g. for the debug link in the manager-bundle.

I disagree. The manager bundle uses an event listener, because that is the only way for other bundles to add menu links.

There is no point in using separate event listeners except for keeping the constructor arguments the way they are now. Other than that, it is just more complicated and more confusing for anyone who needs to work with it.

**Comment by aschempp** [comment]
I kinda agree with @Toflar. The builder is responsible for providing an event to build the menu, not to actually build the menu. That's what the event listeners are for. If the core menu items are not added by an event listener, I can't use priorities to be before/after that listener.

File: core-bundle/src/EventListener/BackendMenuListener.php
Comment (aschempp) [reply core-bundle/src/EventListener/BackendMenuListener.php]: oh well… the diff showed this as the old code, must have gotten me confused 😂

**Review by aschempp — CHANGES_REQUESTED** [review]
Still standing by my statement (this is to shut up PR reminders 😂 )

**Comment by leofeyer** [comment]
I have extracted the main menu changes into a separate PR: #1101

**Comment by leofeyer** [comment]
The PR has been rebased and changed according to what we have discussed. Please review again. 😊

**Review by Toflar — DISMISSED** [review]
So much cleaner now, with proper responsibilities! I like this 😄 👍 
Should have more than one PR approval here before merge though.

File: core-bundle/src/EventListener/BackendMenuListener.php
Comment (aschempp) [inline core-bundle/src/EventListener/BackendMenuListener.php]: The button should have a label or title, so we should probably not remove `MSC.burgerTitle`

File: core-bundle/src/Resources/views/Backend/be_header_menu.html.twig
Comment (aschempp) [inline core-bundle/src/Resources/views/Backend/be_header_menu.html.twig:3]: don't you need to add the translation domain here?

**Review by aschempp — COMMENTED** [review]
Generally looks good. I think there's one optimization possible: 
Same as the debug button, the preview and logout button should be added in their own menu listener. These service listener should contain the code from `PreviewUrlGenerator` and `LogoutUrlGenerator`, because the url generators don't make much sense to me. They're never being reused, they are solely for the purpose of generating that button. It would also make DI for the regular menu builder much simpler, because it wouldn't contain multiple services only needed for one button each.

File: core-bundle/src/Resources/views/Backend/be_header_menu.html.twig
Comment (leofeyer) [reply core-bundle/src/Resources/views/Backend/be_header_menu.html.twig:3]: There is a `trans_default_domain 'contao_default'` at the top. 😉

File: core-bundle/src/EventListener/BackendMenuListener.php
Comment (leofeyer) [reply core-bundle/src/EventListener/BackendMenuListener.php]: I don't think so, because the button is only visible on mobile devices and neither title nor label would be shown.

**Comment by leofeyer** [comment]
>  It would also make DI for the regular menu builder much simpler, because it wouldn't contain multiple services only needed for one button each.

The services do not generate the buttons but only their URLs. This means whenever you want to generate the logout URL or the preview URL, you can use the service.

As soon as #989 is merged, we can reuse the preview URL service in the `tl_page` DCA to generate the URLs for the direct links (when you click the icon on the left).

File: core-bundle/src/EventListener/BackendMenuListener.php
Comment (aschempp) [reply core-bundle/src/EventListener/BackendMenuListener.php]: I would assume it is visible to screen readers, and most likely to an iPhone screen reader as well

**Comment by aschempp** [comment]
> This means whenever you want to generate the logout URL or the preview URL, you can use the service.
> 
> As soon as #989 is merged, we can reuse the preview URL service in the `tl_page` DCA to generate the URLs for the direct links (when you click the icon on the left).

No you cannot. The service is explicitly bound to the `CURRENT_ID` and only works for exactly that button use case. That's what I mean, there's never a case for using that service standalone. You can't pass the current page ID or thelike.

**Comment by leofeyer** [comment]
I have implemented the requested changes. Please review again.

**Review by Toflar — DISMISSED** [review]

File: core-bundle/src/EventListener/Menu/BackendLogoutListener.php
Comment (aschempp) [inline core-bundle/src/EventListener/Menu/BackendLogoutListener.php]: what is this user needed for?

File: core-bundle/src/EventListener/Menu/BackendLogoutListener.php
Comment (aschempp) [inline core-bundle/src/EventListener/Menu/BackendLogoutListener.php]: the variable assignment isn't really necessary, it's not used again but 🤷‍♂

File: core-bundle/src/EventListener/Menu/BackendLogoutListener.php
Comment (aschempp) [inline core-bundle/src/EventListener/Menu/BackendLogoutListener.php]: You don't need `sprintf` here, the translator does support parameters.
```php
$this->translator->trans('MSC.switchBT', [$token->getOriginalToken()->getUsername()], 'contao_default');
```

File: manager-bundle/src/DependencyInjection/Configuration.php
Comment (aschempp) [inline manager-bundle/src/DependencyInjection/Configuration.php]: is this change related to the PR?

**Review by aschempp — COMMENTED** [review]
Looks very good now to me! Just a few minor notes 👍

File: core-bundle/src/EventListener/Menu/BackendLogoutListener.php
Comment (leofeyer) [reply core-bundle/src/EventListener/Menu/BackendLogoutListener.php]: To bypass the listener if the user is not a back end user.

File: core-bundle/src/EventListener/Menu/BackendLogoutListener.php
Comment (leofeyer) [reply core-bundle/src/EventListener/Menu/BackendLogoutListener.php]: You mean it should be moved to line 67? If so, I agree. 😄

Comment (leofeyer) [reply core-bundle/src/EventListener/Menu/BackendLogoutListener.php]: Changed in 1bf08f8fc15d2241edda342e57602b1b59a285e2.

File: core-bundle/src/EventListener/Menu/BackendLogoutListener.php
Comment (leofeyer) [reply core-bundle/src/EventListener/Menu/BackendLogoutListener.php]: Changed in 1bf08f8fc15d2241edda342e57602b1b59a285e2.

File: manager-bundle/src/DependencyInjection/Configuration.php
Comment (leofeyer) [reply manager-bundle/src/DependencyInjection/Configuration.php]: Nope. Reverted in 1bf08f8fc15d2241edda342e57602b1b59a285e2.

File: core-bundle/src/EventListener/Menu/BackendLogoutListener.php
Comment (leofeyer) [reply core-bundle/src/EventListener/Menu/BackendLogoutListener.php]: Switched to checking the user role in 2e36141804f864230ccbc9ca1152972e79df494c.

**Review by aschempp — APPROVED** [review]

**Pull request merged**
